sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs

```diff
@@ -92,31 +93,62 @@ private async Task<int> WhileRunningAsync()
                     {
                         if (_inFlightTasks.Count == 0)
                         {
+                            // No more tasks in flight: exit the loop.
                             break;
                         }
 
                         // Grab all the tasks we currently have running.
                         tasks.AddRange(_inFlightTasks.Keys);
                     }
 
-                    // Now, wait for one of them to finish.
-                    var task = await Task.WhenAny(tasks).ConfigureAwait(false);
-                    List<string> descriptions;
-                    lock (_inFlightTasks)
+                    // Wait for one of the two events to happen:
+                    // 1. All tasks in the list complete successfully, or
+                    // 2. Any task throws an exception.
+                    // There's no standard API with this semantics, so we create a custom completion source that is
+                    // completed when remaining count is zero, or when an exception is thrown.
+                    var remaining = tasks.Count;
+                    var tcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    tasks.ForEach(HandleCompletion);
+                    async void HandleCompletion(Task task)
                     {
-                        // Once finished, remove it from the set of tasks that are running.
-                        descriptions = _inFlightTasks[task];
-                        _inFlightTasks.Remove(task);
+                        try
+                        {
+                            // Wait for the task completion.
+                            await task.ConfigureAwait(false);
+
+                            // Log the descriptions of completed tasks.
+                            var descriptions = _inFlightTasks[task];
+                            foreach (var description in descriptions)
+                                Serilog.Log.Information($"Completed task: {description}");
```

Comment:
This needs braces, but do we need this logging at all, or was this left over from debugging?

```suggestion
                            foreach (var description in descriptions)
                            {
                                Serilog.Log.Information($"Completed task: {description}");
                            }
```

Reply:
Yeah, we have a bunch of places where we don't use braces for one-liners. We could standardize on always-braces, although sometimes it gets lengthy.
I believe all serilog calls are "left over from debugging". They do nothing as we don't initialize Serilog sinks unless this line is uncommented https://github.com/pulumi/pulumi/blob/653dcf8f1fe536916f1dd2b0e294de9173f38461/sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs#L152

Reply:
Okay, got it. As long as they're latent then I'm fine leaving them in.

Though I agree that always-braces can be lengthy, I do think we should standardize on that style.